File: Widget/app/today_view_controller.rb
class TodayViewController < UIViewController
  extend IB

  outlet :clockImageView, UIImageView
  outlet :periodNameLabel, UILabel
  outlet :timeRemainingLabel, UILabel

  def viewDidLoad
    super

    FkP.setup

    @periodNameLabel.text = " "
    @timeRemainingLabel.text = " "
  end

  def viewDidAppear(animated)
    super

    if FkP.initialSetupComplete?
      @tick = NSTimer.scheduledTimerWithTimeInterval(1, target: self, selector: "refresh", userInfo: nil, repeats: true) if @tick.nil?
    else
      nothing_scheduled
    end
  end

  def viewWillDisappear(animated)
    super

    unless @tick.nil?
      @tick.invalidate
      @tick = nil
    end
  end

  def didReceiveMemoryWarning
    super
    # Dispose of any resources that can be recreated.
  end

  def refresh
    status = FkP.status(@clockImageView.bounds)
    @clockImageView.image = status[:clock]
    @periodNameLabel.text = status[:periodName]
    @timeRemainingLabel.text = status[:timeRemaining]
  end

  def nothing_scheduled
    @clockImageView.image = Clock.blank(@clockRect)
    @periodNameLabel.text = "Free time"
    @timeRemainingLabel.text = "Nothing scheduled"
  end

  def widgetPerformUpdateWithCompletionHandler(completionHandler)
    FkP.setup

    if FkP.initialSetupComplete?
      refresh
    else
      nothing_scheduled
    end

    completionHandler.call(NCUpdateResultNewData)
  end

  def widgetMarginInsetsForProposedMarginInsets(defaultMarginInsets)
    # defaults: top: 0, left: 47, bottom: 39, right: 0)
    UIEdgeInsetsMake(10, defaultMarginInsets.left, 10, 0)
  end

end

File: Widget/TodayViewController.swift
import UIKit
import NotificationCenter
import KingPastaKit

class TodayViewController: UIViewController, NCWidgetProviding {
    
    @IBOutlet weak var clockImageView: UIImageView!
    @IBOutlet weak var periodNameLabel: UILabel!
    @IBOutlet weak var timeRemainingLabel: UILabel!
    
    private var tick: NSTimer?
        
    override func viewDidLoad() {
        super.viewDidLoad()
        
        periodNameLabel.text = " "
        timeRemainingLabel.text = " "
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        // FIXME: Called on disapper?
        print("did appear")
        if FkP.isInitialSetupComplete {
            if tick == nil {
                tick = NSTimer.scheduledTimerWithTimeInterval(1, target: self, selector: "refresh", userInfo: nil, repeats: true)
            }
        } else {
            nothingScheduled()
        }
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        
        if let tick = tick {
            tick.invalidate()
            self.tick = nil
        }
    }
    
    func refresh() {
        print("tick!")
        let status = FkP.status(clockImageView.bounds)
        clockImageView.image = status.clock
        periodNameLabel.text = status.periodName
        timeRemainingLabel.text = status.timeRemaining
    }
    
    func nothingScheduled() {
        clockImageView.image = Clock.blank(clockImageView.bounds)
        periodNameLabel.text = "Free time"
        timeRemainingLabel.text = "Nothing scheduled"
    }
    
    func widgetPerformUpdateWithCompletionHandler(completionHandler: ((NCUpdateResult) -> Void)) {
        if FkP.isInitialSetupComplete {
            refresh()
        } else {
         nothingScheduled()
        }

        completionHandler(NCUpdateResult.NewData)
    }
    
    func widgetMarginInsetsForProposedMarginInsets(defaultMarginInsets: UIEdgeInsets) -> UIEdgeInsets {
        // defaults: top: 0, left: 47, bottom: 39, right: 0)
        return UIEdgeInsets(top: 10, left: defaultMarginInsets.left, bottom: 10, right: 0)
    }
    
}
